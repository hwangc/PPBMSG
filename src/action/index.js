import {
  types,
  GEO_API,
  initialStateMap,
  initialStateOrder
} from '../constant';
import {
  getItemsInfo,
  getAddressFromOrder,
  getItemIds,
  getLatLng,
  getOrderFetchUrl,
  getItemFetchUrl,
  getLocationFetchUrl,
  getLocationFetchHeaders
} from '../lib/common';

export const startFetchOrderAction = () => {
  return {
    type: types.START_FETCH_ORDER,
    payload: initialStateOrder
  };
};

export const successFetchOrderAction = order => {
  const orderID = order.order_no;
  const prdNum = order.product ? order.product.length : 0;
  const orderAddress = getAddressFromOrder(order);
  const itemsInfo = order.itemsInfo;

  return {
    type: types.SUCCESS_FETCH_ORDER,
    payload: {
      id: orderID,
      prdNum,
      orderAddress,
      itemsInfo,
      error: ''
    }
  };
};

export const failFetchOrderAction = error => {
  return {
    type: types.FAIL_FETCH_ORDER,
    payload: initialStateOrder
  };
};

export const successFetchLocAction = location => {
  return {
    type: types.SUCCESS_FETCH_LOC,
    payload: {
      location
    }
  };
};

export const failFetchLocAction = error => {
  return {
    type: types.FAIL_FETCH_LOC,
    payload: initialStateMap
  };
};

export const fetchOrder = mall => async dispatch => {
  dispatch(startFetchOrderAction());
  try {
    const resOrder = await fetch(getOrderFetchUrl(mall));
    const resOrderJson = await resOrder.json();
    let order = resOrderJson.response.result;
    const itemIds = getItemIds(order);
    const resItem = await fetch(getItemFetchUrl(itemIds, mall));
    const resItemJson = await resItem.json();
    let item = resItemJson.response.result;
    const itemsInfo = getItemsInfo(item);
    const orderWithItem = Object.assign(order[0], {
      itemsInfo
    });
    const resLatLng = await fetch(
      getLocationFetchUrl(GEO_API.DAUM, orderWithItem),
      getLocationFetchHeaders(GEO_API.DAUM)
    );
    const resLatLngJson = await resLatLng.json();
    let location = getLatLng(GEO_API.DAUM, resLatLngJson);
    dispatch(successFetchOrderAction(orderWithItem));
    dispatch(successFetchLocAction(location));
  } catch (error) {
    dispatch(failFetchOrderAction(error));
    dispatch(failFetchLocAction(error));
  }
};
