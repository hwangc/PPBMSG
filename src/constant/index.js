export const SUCCESS_FETCH_ORDER = 'SUCCESS_FETCH_ORDER';
export const FAIL_FETCH_ORDER = 'FAIL_FETCH_ORDER';
export const START_FETCH_ORDER = 'START_FETCH_ORDER';
export const SUCCESS_FETCH_LOC = 'SUCCESS_FETCH_LOC';
export const FAIL_FETCH_LOC = 'FAIL_FETCH_LOC';
export const START_FETCH_LOC = 'START_FETCH_LOC';
export const types = {
  START_FETCH_ORDER,
  SUCCESS_FETCH_ORDER,
  FAIL_FETCH_ORDER,

  START_FETCH_LOC,
  SUCCESS_FETCH_LOC,
  FAIL_FETCH_LOC
};
export const CHUU = 'chuu';
export const MOSSBEAN = 'mossbean';
export const ICECREAM12 = 'icecream12';
export const MALL = {
  CHUU,
  MOSSBEAN,
  ICECREAM12
};
export const initialStateOrder = {
  id: null,
  prdNum: null,
  orderAddress: null,
  itemsInfo: [],
  error: null
};

export const initialStateMap = {
  latitude: 0,
  longitude: 0,
  latitudeDelta: 0.0922,
  longitudeDelta: 0.0421
};

export const ORDER = {
  order_no: 'o_1',
  product: [
    { product_no: 'p_1' },
    { product_no: 'p_2' },
    { product_no: 'p_3' }
  ],
  ship_address1: 'hello',
  ship_address2: 'address'
};
const GOOGLE = 'google';
const DAUM = 'daum';
const FREEGEOIP = 'freegeoip';
export const GEO_API = {
  GOOGLE,
  DAUM,
  FREEGEOIP
};
