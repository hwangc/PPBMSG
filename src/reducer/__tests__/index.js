import mockStore from 'redux-mock-store';
import * as actions from '../../action';
import { ppbMsg, initialState } from '../index';
import { ORDER } from '../../constant';

const store = mockStore();

beforeEach(() => {
  store.clearActions();
});

describe('PPB Reducer', () => {
  test('Success to fetch and dispatch ADD_NEW_ORDER', () => {
    expect(
      ppbMsg(initialState, actions.successFetchAction(ORDER))
    ).toMatchSnapshot();
  });
});
