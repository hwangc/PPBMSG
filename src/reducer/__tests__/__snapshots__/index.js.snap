// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PPB Reducer Success to fetch and dispatch ADD_NEW_ORDER 1`] = `
Object {
  "error": "",
  "id": undefined,
  "images": undefined,
  "orderAddress": "undefined undefined",
  "prdNum": 0,
}
`;
